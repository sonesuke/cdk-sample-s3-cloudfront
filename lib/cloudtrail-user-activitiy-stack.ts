import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as cloudtrail from 'aws-cdk-lib/aws-cloudtrail';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as iam from 'aws-cdk-lib/aws-iam';

export class CloudTrailUserActivityStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const logBucket = scope.node.tryGetContext('cloudtrail-log-bucket');

    // S3にCloudTrailのログを保存するためのバケットを作成
    const bucket = new s3.Bucket(this, 'app-cloudtrail-bucket', {
      bucketName: logBucket,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      removalPolicy: cdk.RemovalPolicy.RETAIN,
    });

    // CloudTrailの設定
    const trail = new cloudtrail.Trail(this, 'app-cloudtrail', {
      sendToCloudWatchLogs: true,
      cloudWatchLogsRetention: logs.RetentionDays.ONE_MONTH,
      bucket: bucket,
      isMultiRegionTrail: true,
    });

    // CloudTrailのログをS3に保存するための権限を付与
    bucket.addToResourcePolicy(
      new iam.PolicyStatement({
        actions: ['s3:GetBucketAcl', 's3:PutObject'],
        resources: [bucket.bucketArn],
        principals: [new iam.ServicePrincipal('cloudtrail.amazonaws.com')],
        effect: iam.Effect.ALLOW,
      })
    );
  }
}
