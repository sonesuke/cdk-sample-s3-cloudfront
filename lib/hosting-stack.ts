import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';

export interface HostingStackProps extends cdk.StackProps {
  bucket: s3.IBucket;
}

export class HostingStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: HostingStackProps) {
    super(scope, id, props);

    // CloudFrontの設定
    const distribution = new cloudfront.CloudFrontWebDistribution(
      this,
      'app-cloudfront-distribution',
      {
        originConfigs: [
          {
            s3OriginSource: {
              // S3をオリジンとして指定
              s3BucketSource: props.bucket,
            },
            behaviors: [{ isDefaultBehavior: true }], // デフォルトの振る舞いを指定
          },
        ],
      }
    );
  }
}
