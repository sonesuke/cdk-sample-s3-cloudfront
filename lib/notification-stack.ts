import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subs from 'aws-cdk-lib/aws-sns-subscriptions';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as guardduty from 'aws-cdk-lib/aws-guardduty';

interface NotificationStackProps extends cdk.StackProps {
  bucket: s3.IBucket;
}

export class NotificationStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: NotificationStackProps) {
    super(scope, id, props);

    const mailAddress = scope.node.tryGetContext('notification-mail') as string;

    if (!mailAddress) {
      throw new Error('mailAddress is not defined in CDK context');
    }

    // 通知を受け取るSNSトピックを作成
    const topic = new sns.Topic(this, 'app-notification-sns-topic');

    // 通知先をメールアドレスに設定
    topic.addSubscription(new subs.EmailSubscription(mailAddress));

    // S3のPutObjectイベントをトリガーにEventBridgeルールを作成
    const uploadEvent = new events.Rule(this, 'app-notification-rule-upload-event', {
      eventPattern: {
        source: ['aws.s3'],
        detailType: ['AWS API Call via CloudTrail'],
        detail: {
          eventSource: ['s3.amazonaws.com'],
          eventName: ['PutObject', 'CopyObject', 'CompleteMultipartUpload'],
          requestParameters: {
            bucketName: [props.bucket.bucketName],
          },
        },
      },
    });

    // ルールとSNSトピックを紐付け
    uploadEvent.addTarget(new targets.SnsTopic(topic));

    // GuardDutyの有効化
    const guardduty_enable = new guardduty.CfnDetector(this, 'app-notificiation-guardduty', {
      enable: true,
    });

    // GuardDutyのイベントをトリガーにEventBridgeルールを作成
    const guarddutyRule = new events.Rule(this, 'app-notificaition-rule-guardduty', {
      eventPattern: {
        source: ['aws.guardduty'],
        detailType: ['GuardDuty Finding'],
        detail: {
          severity: [
            { numeric: ['>=', 4] }, // 重要度中・高のみ通知するようにフィルタリング
          ],
        },
      },
    });

    // GuardDutyのルールとSNSトピックを紐付け
    guarddutyRule.addTarget(new targets.SnsTopic(topic));
  }
}
