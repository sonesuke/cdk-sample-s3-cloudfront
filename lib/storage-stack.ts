import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as s3 from 'aws-cdk-lib/aws-s3';

export class StorageStack extends cdk.Stack {
  bucket: s3.IBucket;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.bucket = new s3.Bucket(this, 'app-storage-bucket', {
      bucketName: 'sample-bucket-202303260926-jst',
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      removalPolicy: cdk.RemovalPolicy.RETAIN,
      versioned: true,
      encryption: s3.BucketEncryption.S3_MANAGED,
    });
  }
}
