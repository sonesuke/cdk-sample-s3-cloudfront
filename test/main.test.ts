import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CloudTrailUserActivityStack } from '../lib/cloudtrail-user-activitiy-stack';
import { HostingStack } from '../lib/hosting-stack';
import { NotificationStack } from '../lib/notification-stack';
import { StorageStack } from '../lib/storage-stack';

test('All Stacks created', () => {
  const app = new cdk.App({
    context: {
      'hosting-bucket': 'sample-hosting-bucket',
      'cloudtrail-log-bucket': 'sample-log-bucket',
      'notification-mail': 'example@sample.com',
    },
  });

  const cloudtrail = new CloudTrailUserActivityStack(app, 'app-stack-cloudtrail-user-activity');

  const storage = new StorageStack(app, 'app-stack-storage');

  const hosting = new HostingStack(app, 'app-stack-hosting', { bucket: storage.bucket });

  const notificaiton = new NotificationStack(app, 'app-stack-notification', {
    bucket: storage.bucket,
  });

  expect(Template.fromStack(cloudtrail).toJSON()).toMatchSnapshot();
  expect(Template.fromStack(storage).toJSON()).toMatchSnapshot();
  expect(Template.fromStack(hosting).toJSON()).toMatchSnapshot();
  expect(Template.fromStack(notificaiton).toJSON()).toMatchSnapshot();
});
