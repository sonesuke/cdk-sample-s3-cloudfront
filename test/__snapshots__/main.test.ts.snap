// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All Stacks created 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "appcloudtrailCBC1845F": {
      "DependsOn": [
        "appcloudtrailbucketPolicy603768EC",
        "appcloudtrailLogsRoleDefaultPolicy0B77D06D",
        "appcloudtrailLogsRole6DFA92EF",
      ],
      "Properties": {
        "CloudWatchLogsLogGroupArn": {
          "Fn::GetAtt": [
            "appcloudtrailLogGroupAFF56743",
            "Arn",
          ],
        },
        "CloudWatchLogsRoleArn": {
          "Fn::GetAtt": [
            "appcloudtrailLogsRole6DFA92EF",
            "Arn",
          ],
        },
        "EnableLogFileValidation": true,
        "EventSelectors": [],
        "IncludeGlobalServiceEvents": true,
        "IsLogging": true,
        "IsMultiRegionTrail": true,
        "S3BucketName": {
          "Ref": "appcloudtrailbucket7C4AF267",
        },
      },
      "Type": "AWS::CloudTrail::Trail",
    },
    "appcloudtrailLogGroupAFF56743": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RetentionInDays": 30,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "appcloudtrailLogsRole6DFA92EF": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudtrail.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "appcloudtrailLogsRoleDefaultPolicy0B77D06D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutLogEvents",
                "logs:CreateLogStream",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "appcloudtrailLogGroupAFF56743",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "appcloudtrailLogsRoleDefaultPolicy0B77D06D",
        "Roles": [
          {
            "Ref": "appcloudtrailLogsRole6DFA92EF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "appcloudtrailbucket7C4AF267": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "sample-log-bucket",
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "appcloudtrailbucketPolicy603768EC": {
      "Properties": {
        "Bucket": {
          "Ref": "appcloudtrailbucket7C4AF267",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetBucketAcl",
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudtrail.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "appcloudtrailbucket7C4AF267",
                  "Arn",
                ],
              },
            },
            {
              "Action": "s3:PutObject",
              "Condition": {
                "StringEquals": {
                  "s3:x-amz-acl": "bucket-owner-full-control",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudtrail.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "appcloudtrailbucket7C4AF267",
                        "Arn",
                      ],
                    },
                    "/AWSLogs/",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    "/*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "s3:GetBucketAcl",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudtrail.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "appcloudtrailbucket7C4AF267",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`All Stacks created 2`] = `
{
  "Outputs": {
    "ExportsOutputFnGetAttappstoragebucket4C9F4AE1RegionalDomainNameE4668BAF": {
      "Export": {
        "Name": "app-stack-storage:ExportsOutputFnGetAttappstoragebucket4C9F4AE1RegionalDomainNameE4668BAF",
      },
      "Value": {
        "Fn::GetAtt": [
          "appstoragebucket4C9F4AE1",
          "RegionalDomainName",
        ],
      },
    },
    "ExportsOutputRefappstoragebucket4C9F4AE1FF4B51D6": {
      "Export": {
        "Name": "app-stack-storage:ExportsOutputRefappstoragebucket4C9F4AE1FF4B51D6",
      },
      "Value": {
        "Ref": "appstoragebucket4C9F4AE1",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "appstoragebucket4C9F4AE1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "sample-bucket-202303260926-jst",
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`All Stacks created 3`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "appcloudfrontdistributionCFDistribution72D2ECCD": {
      "Properties": {
        "DistributionConfig": {
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
            ],
            "CachedMethods": [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "DomainName": {
                "Fn::ImportValue": "app-stack-storage:ExportsOutputFnGetAttappstoragebucket4C9F4AE1RegionalDomainNameE4668BAF",
              },
              "Id": "origin1",
              "S3OriginConfig": {},
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": {
            "CloudFrontDefaultCertificate": true,
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`All Stacks created 4`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "appnotificaitionruleguardduty9D23A371": {
      "Properties": {
        "EventPattern": {
          "detail": {
            "severity": [
              {
                "numeric": [
                  ">=",
                  4,
                ],
              },
            ],
          },
          "detail-type": [
            "GuardDuty Finding",
          ],
          "source": [
            "aws.guardduty",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Ref": "appnotificationsnstopic87519ED5",
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "appnotificationruleuploadeventF71A2B4D": {
      "Properties": {
        "EventPattern": {
          "detail": {
            "eventName": [
              "PutObject",
              "CopyObject",
              "CompleteMultipartUpload",
            ],
            "eventSource": [
              "s3.amazonaws.com",
            ],
            "requestParameters": {
              "bucketName": [
                {
                  "Fn::ImportValue": "app-stack-storage:ExportsOutputRefappstoragebucket4C9F4AE1FF4B51D6",
                },
              ],
            },
          },
          "detail-type": [
            "AWS API Call via CloudTrail",
          ],
          "source": [
            "aws.s3",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Ref": "appnotificationsnstopic87519ED5",
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "appnotificationsnstopic87519ED5": {
      "Type": "AWS::SNS::Topic",
    },
    "appnotificationsnstopicPolicyD8639FD0": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
              "Resource": {
                "Ref": "appnotificationsnstopic87519ED5",
              },
              "Sid": "0",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": [
          {
            "Ref": "appnotificationsnstopic87519ED5",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
    "appnotificationsnstopicexamplesamplecomB061B6D8": {
      "Properties": {
        "Endpoint": "example@sample.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "appnotificationsnstopic87519ED5",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "appnotificiationguardduty": {
      "Properties": {
        "Enable": true,
      },
      "Type": "AWS::GuardDuty::Detector",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
